File: app/schemas/writing_assistance.py
"""Pydantic schemas for Writing Assistance API."""
from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict

from app.models.writing_assistance import WritingFeedbackType


# ---------------------------------------------------------------------------
# Request schemas
# ---------------------------------------------------------------------------


class WritingAnalysisRequest(BaseModel):
    """Request body for analyzing a piece of writing."""

    title: str
    text: str
    course_id: Optional[int] = None
    assignment_type: Optional[str] = "essay"  # essay | report | letter | lab


class WritingImproveRequest(BaseModel):
    """Request body for applying a specific improvement instruction to a session."""

    session_id: int
    instruction: str  # e.g. "make it more formal", "add more evidence"


# ---------------------------------------------------------------------------
# Response schemas
# ---------------------------------------------------------------------------


class WritingFeedbackItem(BaseModel):
    """A single feedback item returned by the AI."""

    type: WritingFeedbackType
    message: str
    suggestion: str
    severity: Literal["info", "warning", "error"]


class WritingAnalysisResponse(BaseModel):
    """Response returned after analyzing a piece of writing."""

    session_id: int
    overall_score: int
    feedback: list[WritingFeedbackItem]
    improved_text: str
    suggestions_count: int
    word_count: int


class WritingImproveResponse(BaseModel):
    """Response returned after applying an improvement instruction."""

    improved_text: str
    instruction: str


class WritingSessionSummary(BaseModel):
    """Summary of a writing session (no full text) for list views."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    assignment_type: str
    overall_score: Optional[int]
    word_count: int
    suggestions_count: Optional[int] = None
    created_at: datetime
    updated_at: Optional[datetime]


class WritingSessionDetail(BaseModel):
    """Full writing session including original and improved text."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    assignment_type: str
    original_text: str
    improved_text: Optional[str]
    feedback: Optional[list[WritingFeedbackItem]]
    overall_score: Optional[int]
    word_count: int
    course_id: Optional[int]
    created_at: datetime
    updated_at: Optional[datetime]


class WritingTemplateResponse(BaseModel):
    """Response schema for a writing template."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    description: str
    template_type: str
    structure_outline: str
    is_active: bool

File: frontend/src/api/writingAssistance.ts
/**
 * Writing Assistance API client.
 *
 * All endpoints require an authenticated user.
 */
import { api } from './client';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export type WritingFeedbackType =
  | 'grammar'
  | 'clarity'
  | 'structure'
  | 'argumentation'
  | 'vocabulary'
  | 'overall';

export type FeedbackSeverity = 'info' | 'warning' | 'error';

export type AssignmentType = 'essay' | 'report' | 'letter' | 'lab';

export interface WritingFeedbackItem {
  type: WritingFeedbackType;
  message: string;
  suggestion: string;
  severity: FeedbackSeverity;
}

export interface WritingAnalysisRequest {
  title: string;
  text: string;
  course_id?: number | null;
  assignment_type?: AssignmentType;
}

export interface WritingAnalysisResponse {
  session_id: number;
  overall_score: number;
  feedback: WritingFeedbackItem[];
  improved_text: string;
  suggestions_count: number;
  word_count: number;
}

export interface WritingImproveRequest {
  session_id: number;
  instruction: string;
}

export interface WritingImproveResponse {
  improved_text: string;
  instruction: string;
}

export interface WritingSessionSummary {
  id: number;
  title: string;
  assignment_type: string;
  overall_score: number | null;
  word_count: number;
  suggestions_count: number | null;
  created_at: string;
  updated_at: string | null;
}

export interface WritingSessionDetail {
  id: number;
  title: string;
  assignment_type: string;
  original_text: string;
  improved_text: string | null;
  feedback: WritingFeedbackItem[] | null;
  overall_score: number | null;
  word_count: number;
  course_id: number | null;
  created_at: string;
  updated_at: string | null;
}

export interface WritingTemplate {
  id: number;
  name: string;
  description: string;
  template_type: string;
  structure_outline: string;
  is_active: boolean;
}

// ---------------------------------------------------------------------------
// API functions
// ---------------------------------------------------------------------------

/** Analyze a piece of writing and receive scored feedback + improved version. */
export async function analyzeWriting(
  request: WritingAnalysisRequest,
): Promise<WritingAnalysisResponse> {
  const { data } = await api.post<WritingAnalysisResponse>('/api/writing/analyze', request);
  return data;
}

/** Apply a specific improvement instruction to a previous writing session. */
export async function improveWriting(
  request: WritingImproveRequest,
): Promise<WritingImproveResponse> {
  const { data } = await api.post<WritingImproveResponse>('/api/writing/improve', request);
  return data;
}

/** List all writing sessions for the current user (summary only). */
export async function getSessions(): Promise<WritingSessionSummary[]> {
  const { data } = await api.get<WritingSessionSummary[]>('/api/writing/sessions');
  return data;
}

/** Get the full writing session including text and feedback. */
export async function getSession(sessionId: number): Promise<WritingSessionDetail> {
  const { data } = await api.get<WritingSessionDetail>(`/api/writing/sessions/${sessionId}`);
  return data;
}

/** List all active writing templates. */
export async function getTemplates(): Promise<WritingTemplate[]> {
  const { data } = await api.get<WritingTemplate[]>('/api/writing/templates');
  return data;
}
